spec/02-integration/05-proxy/18-upstream_tls_spec.lua

```diff
@@ -125,26 +129,93 @@ for _, strategy in helpers.each_strategy() do
         paths = { "/mtls-upstream", },
       }))
 
+      -- tls
+      tls_service_mtls = assert(bp.services:insert({
+        name = "tls-protected-service-mtls",
+        url = "tls://127.0.0.1:16798",
+      }))
+
+      tls_service_tls = assert(bp.services:insert({
+        name = "tls-protected-service",
+        url = "tls://example.com:16799", -- domain name needed for hostname check
+      }))
+
+      tls_upstream = assert(bp.upstreams:insert({
+        name = "tls-backend-mtls",
+      }))
+
+      assert(bp.targets:insert({
+        upstream = { id = tls_upstream.id, },
+        target = "example.com:16798",
+      }))
+
+      tls_service_mtls_upstream = assert(bp.services:insert({
+        name = "tls-protected-service-mtls-upstream",
+        url = "tls://tls-backend-mtls",
+        host = "example.com"
+      }))
+
+      assert(bp.routes:insert({
+        service = { id = tls_service_mtls.id, },
+        destinations = {
+          {
+            port = 19000,
+          },
+        },
+        protocols = {
+          "tls",
+        },
+      }))
+
+      assert(bp.routes:insert({
+        service = { id = tls_service_tls.id, },
+        destinations = {
+          {
+            port = 19001,
+          },
+        },
+        protocols = {
+          "tls",
+        },
+      }))
+
+      assert(bp.routes:insert({
+        service = { id = tls_service_mtls_upstream.id, },
+        destinations = {
+          {
+            port = 19002,
+          },
+        },
+        protocols = {
+          "tls",
+        },
+      }))
+
+
       assert(helpers.start_kong({
         database   = strategy,
         nginx_conf = "spec/fixtures/custom_nginx.template",
+        stream_listen = helpers.get_proxy_ip(false) .. ":19000," ..
+        helpers.get_proxy_ip(false) .. ":19001," ..
+        helpers.get_proxy_ip(false) .. ":19002"
       }, nil, nil, fixtures))
 
-      proxy_client = assert(helpers.proxy_client())
       admin_client = assert(helpers.admin_client())
     end)
 
     lazy_teardown(function()
-      if proxy_client then
-        proxy_client:close()
-      end
-
       helpers.stop_kong()
     end)
-
-    describe("mutual TLS authentication against upstream with Service object", function()
+    for _, subsystems in pairs({"http", "stream"}) do
+    describe(subsystems .. " mutual TLS authentication against upstream with Service object", function()
       describe("no client certificate supplied", function()
         it("accessing protected upstream", function()
+          local proxy_client
```

Comment:
I would suggest putting some of these bits of logic into a central object with helper functions/properties:

```lua
local SUBSYSTEMS = {
  http = {
    proxy_client = function()
      return helpers.proxy_client()
    end,

    path = "/mtls",
    host = "tls-http.test",

    service_name = "tls-http",
  },

  stream = {
    proxy_client = function()
      return helpers.proxy_client(6000, 19000)
    end,

    path = "/",
    host = "tls-stream.test",

    service_name = "tls-stream",
  },
}
```

Now, in your tests you can do this:

```lua
for subsystem, conf in pairs(SUBSYSTEMS) do
  describe(subsystem .. " something something", function()
    it("works, I think...", function()
      -- for client requests
      local client = conf.proxy_client()
      local res = client:get(conf.path, {
        headers = { host = conf.host },
      })

      -- for admin requests
      local res = admin_client:patch("/services/" .. conf.service_name, {
        -- ...
      })
    end)
  end)
end
```

This could be expanded to even further cut down on boilerplate/repeated code if you wish, so use your discretion.

Reply:
I move proxy_client to the test phase because this client will not keep alive in stream mode, so it must reconnect the stream mode test.

Reply:
I have changed my code in this logic, please review it